3p/3p.js

```diff
@@ -211,15 +194,34 @@ export function computeInMasterFrame(global, taskId, work, cb) {
 }
 
 /**
- * Throws an exception if data does not contains a mandatory field.
+ * Validates given data. Throws an exception if the data does not
+ * contains a mandatory field. If called with the optional param
+ * opt_optionalFields, it also validates that the data contains no fields other
+ * than mandatory and optional fields.
+ *
+ * Mandatory fields also accept a string Array as an item. All items in that
+ * array are considered as alternatives to each other. So the validation checks
+ * that the data contains exactly one of those alternatives.
+ *
  * @param {!Object} data
- * @param {!Array<string>} mandatoryFields
+ * @param {!Array<string|Array<string>>} mandatoryFields
+ * @param {Array<string>=} opt_optionalFields
  */
-export function validateDataExists(data, mandatoryFields) {
+export function validateData(data, mandatoryFields, opt_optionalFields) {
+  let allowedFields = opt_optionalFields || [];
   for (let i = 0; i < mandatoryFields.length; i++) {
     const field = mandatoryFields[i];
-    user().assert(data[field],
-        'Missing attribute for %s: %s.', data.type, field);
+    if (Array.isArray(field)) {
+      validateExactlyOne(data, field);
+      allowedFields = allowedFields.concat(field);
+    } else {
+      user().assert(data[field],
+          'Missing attribute for %s: %s.', data.type, field);
+      allowedFields.push(field);
+    }
+  }
+  if (opt_optionalFields) {
```

Comment:
get your point! A bit confusing as optional = null and optional = [] behave very differently.